core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -197,14 +222,29 @@ class ControllerChannelManager(controllerEpoch: () => Int,
       brokerState.requestSendThread.shutdown()
       brokerState.networkClient.close()
       brokerState.messageQueue.clear()
-      metricsGroup.removeMetric(QueueSizeMetricName, brokerMetricTags(brokerState.brokerNode.id))
-      metricsGroup.removeMetric(RequestRateAndQueueTimeMetricName, brokerMetricTags(brokerState.brokerNode.id))
+      removeMetricsForBroker(brokerState)
       brokerStateInfo.remove(brokerState.brokerNode.id)
     } catch {
       case e: Throwable => error("Error while removing broker by the controller", e)
     }
   }
 
+  private def removeMetricsForBroker(brokerState: ControllerBrokerStateInfo): Unit = {
+    val removeTagSet = new java.util.HashSet[java.util.Map[String, String]]()
+    val metricNameWithTag = new MetricNameWithTagData
+    metricNameWithTag.putAll(gaugeMetricNameWithTag)
+    metricNameWithTag.putAll(timerMetricNameWithTag)
+    metricNameWithTag.asScala.foreach(metricNameAndTags => {
+      metricNameAndTags._2.asScala.filter(tag => tag.get(BrokerMetricTagKeyName).equals(brokerState.brokerNode.id.toString))
+        .foreach(metricTag => {
+          metricsGroup.removeMetric(metricNameAndTags._1, metricTag)
+          removeTagSet.add(metricTag)
+        })
+    })
+    gaugeMetricNameWithTag.values().asScala.foreach(tagSet => tagSet.removeAll(removeTagSet))
+    timerMetricNameWithTag.values().asScala.foreach(tagSet => tagSet.removeAll(removeTagSet))
+  }
```

Comment:
I think this could be simplified without having to create new map and copying data.

e.g. you can pass `gaugeMetricNameWithTag` as a parameter to a function which can iterate through the value , filter the matching tags and remove the metrics for that tag. You can use the same function for timerMetricNameWithTag